Detectors/GlobalTracking/src/MatchTOF.cxx

```diff
@@ -838,6 +1091,290 @@ void MatchTOF::doMatching(int sec)
+{
+
+  auto& gasParam = o2::tpc::ParameterGas::Instance();
+  float vdrift = gasParam.DriftV;
+
+  ///< do the real matching per sector
+  mMatchedTracksPairs.clear(); // new sector
+
+  auto& cacheTOF = mTOFClusSectIndexCache[sec]; // array of cached TOF cluster indices for this sector; reminder: they are ordered in time!
+  auto& cacheTrk = mTracksSectIndexCache[sec];  // array of cached tracks indices for this sector; reminder: they are ordered in time!
+  int nTracks = cacheTrk.size(), nTOFCls = cacheTOF.size();
+  LOG(INFO) << "Matching sector " << sec << ": number of tracks: " << nTracks << ", number of TOF clusters: " << nTOFCls;
+  if (!nTracks || !nTOFCls) {
+    return;
+  }
+  int itof0 = 0;                          // starting index in TOF clusters for matching of the track
+  int detId[2][5];                        // at maximum one track can fall in 2 strips during the propagation; the second dimention of the array is the TOF det index
+  float deltaPos[2][3];                   // at maximum one track can fall in 2 strips during the propagation; the second dimention of the array is the residuals
+  o2::track::TrackLTIntegral trkLTInt[2]; // Here we store the integrated track length and time for the (max 2) matched strips
+  int nStepsInsideSameStrip[2] = {0, 0};  // number of propagation steps in the same strip (since we have maximum 2 strips, it has dimention = 2)
+  float deltaPosTemp[3];
+  std::array<float, 3> pos;
+  std::array<float, 3> posBeforeProp;
+  float posFloat[3];
+
+  // prematching for TPC only tracks (identify BC candidate to correct z for TPC track accordingly to v_drift)
+
+  LOG(DEBUG) << "Trying to match %d tracks" << cacheTrk.size();
+  for (int itrk = 0; itrk < cacheTrk.size(); itrk++) {
+    int nStripsCrossedInPropagation = 0; // how many strips were hit during the propagation
+    auto& trackWork = mTracksWork[cacheTrk[itrk]];
+    auto& trefTrk = trackWork.first;
+    auto& intLT = mLTinfos[cacheTrk[itrk]];
+
+    int nBCcand = 1;
+    std::vector<ulong> BCcand;
+    BCcand.reserve(nBCcand);
+
+    int side = (trefTrk.getZ() > 0) ? 1 : -1;
+
+    // look at BC candidates for the track
+    itof0 = 0;
+    double minTrkTime = (trackWork.second.getTimeStamp() - mSigmaTimeCut * trackWork.second.getTimeStampError()) * 1.E6; // minimum time in ps
+    double maxTrkTime = (trackWork.second.getTimeStamp() + mSigmaTimeCut * trackWork.second.getTimeStampError()) * 1.E6; // maximum time in ps
+    for (auto itof = itof0; itof < nTOFCls; itof++) {
+      auto& trefTOF = mTOFClusWork[cacheTOF[itof]];
+
+      if (trefTOF.getTime() < minTrkTime) { // this cluster has a time that is too small for the current track, we will get to the next one
+        itof0 = itof + 1;
+        continue;
+      }
+
+      if (trefTOF.getTime() > maxTrkTime) { // this cluster has a time that is too large for the current track, close loop
+        break;
+      }
+
+      if (trefTOF.getZ() * side < 0)
```

Comment:
not sure this is correct: the point on the TPC track cannot change the side only if it is within the TPC radius, but since the tracks can be moved within the side, its extrapolation to large R may change the Z sign.

Reply:
Is there a way provided by the TPC track class?

Reply:
well, the tpcTrack has methods ``hasASideClustersOnly()``, ``hasCSideClusters()``, ``hasBothSidesClusters()``, but this does not tell you unabmigously what will be the track position at TOF radius: if the eta>0 and it is A side track (or CE-crossing), then the extrapolation will be on A side. Idem for eta<0 and C-side or CE-crossing tracks. But for remaining combinations you should check the extrapolated Z (note that there is "fast" method ``track.getZAt(x, b)`` (then you should assume some margin due to using 1d B-field).

Reply:
Here you will still miss the TPC tracks which change the side after extrapolation to TOF. One could avoid this by
````
if ( (trefTOF.getZ() * side < 0) && ((side>0) != (trackWork.first.getTgl()>0))) {
  continue;
}
````